pkg/auth/auth.go

```diff
@@ -153,3 +161,186 @@ func (c *Client) RevokeToken(ctx context.Context, req *RevokeTokenRequest) error
+		Username: username,
+		Password: password,
+		API:      api,
+	}, nil
+}
+
+// GetAccessToken manages the token state and returns a valid access token
+func (h *Helper) GetAccessToken() (string, error) {
+	h.mutex.Lock()
+	defer h.mutex.Unlock()
+
+	// Check if the tokenstore file exists
+	if _, err := os.Stat(tokenStoreFile); os.IsNotExist(err) {
+		// File does not exist, create and login
+		return h.loginAndStoreTokens(tokenStoreFile)
+	}
+
+	// File exists, read and unmarshal token data
+	data, err := os.ReadFile(tokenStoreFile)
+	if err != nil {
+		return "", err
+	}
+
+	var tokenStore Tokenstore
+	if err := json.Unmarshal(data, &tokenStore); err != nil {
+		return "", err
+	}
+
+	// Check if the access token is still valid
+	if time.Since(tokenStore.AccessTokenGeneratedAt) < 24*time.Hour {
+		return tokenStore.AccessToken, nil
+	}
+
+	// Check if the refresh token is still valid
+	if time.Since(tokenStore.RefreshTokenGeneratedAt) < 30*24*time.Hour {
+		return h.refreshAndStoreTokens(tokenStoreFile, &tokenStore)
+	}
+
+	// Both tokens are expired, login again
+	return h.loginAndStoreTokens(tokenStoreFile)
+}
+
+// loginAndStoreTokens logs in and stores the tokens
+func (h *Helper) loginAndStoreTokens(tokenStoreFile string) (string, error) {
+	loginRequest := &LoginRequest{
+		Username: h.Username,
+		Password: h.Password,
+	}
+
+	loginResponse, err := h.API.Login(context.Background(), loginRequest)
+	if err != nil {
+		return "", err
+	}
+
+	tokenStore := &Tokenstore{
+		AccessToken:             loginResponse.AccessToken,
+		AccessTokenGeneratedAt:  time.Now(),
+		RefreshToken:            loginResponse.RefreshToken,
+		RefreshTokenGeneratedAt: time.Now(),
+	}
+
+	if err := h.storeTokens(tokenStoreFile, tokenStore); err != nil {
+		return "", err
+	}
+
+	return tokenStore.AccessToken, nil
+}
+
+// refreshAndStoreTokens refreshes the access token and stores the new tokens
+func (h *Helper) refreshAndStoreTokens(tokenStoreFile string, tokenStore *Tokenstore) (string, error) {
+	refreshRequest := &RefreshTokenRequest{
+		Username:     h.Username,
+		RefreshToken: tokenStore.RefreshToken,
+	}
+
+	var refreshResponse *RefreshTokenResponse
+	var err error
+
+	// Check if the refresh token is less than 30 days old
+	if time.Since(tokenStore.RefreshTokenGeneratedAt) < 30*24*time.Hour {
+		refreshResponse, err = h.API.RefreshToken(context.Background(), refreshRequest)
+	}
+
+	if err != nil {
+		return "", err
+	}
+
+	tokenStore.AccessToken = refreshResponse.AccessToken
+	tokenStore.AccessTokenGeneratedAt = time.Now()
+
+	if err := h.storeTokens(tokenStoreFile, tokenStore); err != nil {
+		return "", err
+	}
+
+	return tokenStore.AccessToken, nil
+}
+
+// storeTokens writes the token data to the tokenstore file in json
+func (h *Helper) storeTokens(tokenStoreFile string, tokenStore *Tokenstore) error {
+	data, err := json.Marshal(tokenStore)
+	if err != nil {
+		return err
+	}
+
+	//nolint:errcheck
+	return os.WriteFile(tokenStoreFile, data, 0600)
```

Comment:
Should we be doing `mutex` lock while writing?

Reply:
Added a mutex lock